server/src/service.rs

```diff
@@ -123,46 +125,85 @@ impl Service {
 }
 
 impl Service {
-    pub async fn run() -> Result<()> {
+    pub async fn run() -> Result<zbus::Connection> {
         let connection = zbus::connection::Builder::session()?
             .name(oo7::dbus::api::Service::DESTINATION.as_deref().unwrap())?
             .build()
             .await?;
-        let object_server = connection.object_server();
         let service = Self {
             collections: Default::default(),
             manager: Default::default(),
             connection: connection.clone(),
         };
-        let collections = service.fetch_collections().await;
 
-        object_server
+        connection
+            .object_server()
             .at(oo7::dbus::api::Service::PATH.as_deref().unwrap(), service)
             .await?;
 
-        for collection in collections {
-            object_server
-                .at(collection.path().clone(), collection)
+        Ok(connection)
+    }
+
+    pub async fn dispatch_collections(
+        connection: zbus::Connection,
+        secret: Option<Secret>,
+    ) -> Result<()> {
+        let service_ifce_ref = connection
+            .object_server()
+            .interface::<_, Service>(oo7::dbus::api::Service::PATH.as_deref().unwrap())
+            .await?;
+        let service = service_ifce_ref.get_mut().await;
+        let manager = service.manager();
+
+        if secret.is_some() {
+            // create and dispatch login/default collection
+            let login = Collection::new(
+                "login",
+                "default",
+                manager.clone(),
+                Arc::new(match Keyring::open("login", secret.unwrap()).await {
+                    Ok(keyring) => keyring,
+                    Err(err) => {
+                        tracing::error!("Failed to open login keyring: {}", err);
+                        std::process::exit(1);
+                    }
+                }),
+            );
+            service.set_collections(login.path().clone()).await;
+            connection
+                .object_server()
+                .at(login.path().clone(), login)
                 .await?;
         }
 
+        // create and dispatch temporary session collection
+        let secret = Secret::from(rand::thread_rng().gen::<[u8; 8]>().to_vec());
+        let session = Collection::new(
+            "session",
+            "session",
+            manager,
+            Arc::new(match Keyring::temporary(secret).await {
```

Comment:
Don't call std::process::exit here please. just propagate the error by calling Keyring::temporary(secret).await?

Reply:
Done, thanks!